File: LAB4/task_3/linear_congruential_generator.cpp
#include <iostream>
#include <vector>
#include <unordered_map>

using namespace std;

/**
 * @brief Генерирует последовательность чисел с помощью линейного конгруэнтного генератора и находит индекс начала цикла.
 * @param X0 Начальное значение (seed).
 * @param A Множитель.
 * @param B Инкремент.
 * @param C Модуль.
 */
void generateLCG(int X0, int A, int B, int C) {
    unordered_map<int, int> seen; // Хранит число и его индекс
    vector<int> sequence;         // Последовательность чисел
    int current = X0;
    int index = 0;

    // Генерация чисел до обнаружения повтора
    while (true) {
        if (seen.find(current) != seen.end()) {
            // Вывод последовательности
            for (int num : sequence) {
                cout << num << " ";
            }
            // Вывод индекса начала цикла
            cout << "\n" << seen[current] << endl;
            break;
        }
        seen[current] = index;
        sequence.push_back(current);
        current = (static_cast<long long>(A) * current + B) % C;
        index++;
    }
}

int main() {
    int X0, A, B, C;
    // Ввод параметров
    cin >> X0 >> A >> B >> C;
    generateLCG(X0, A, B, C);
    return 0;
}

File: LAB4/task_3/linear_congruential_generator.go
package main

import (
	"fmt"
)

/**
 * @brief Генерирует последовательность чисел с помощью линейного конгруэнтного генератора и находит индекс начала цикла.
 * @param X0 Начальное значение (seed).
 * @param A Множитель.
 * @param B Инкремент.
 * @param C Модуль.
 */
func generateLCG(X0, A, B, C int) {
	seen := make(map[int]int) // Хранит число и его индекс
	sequence := []int{}       // Последовательность чисел
	current := X0
	index := 0

	// Генерация чисел до обнаружения повтора
	for {
		if idx, exists := seen[current]; exists {
			// Вывод последовательности
			for _, num := range sequence {
				fmt.Printf("%d ", num)
			}
			// Вывод индекса начала цикла
			fmt.Printf("\n%d\n", idx)
			break
		}
		seen[current] = index
		sequence = append(sequence, current)
		current = (A*current + B) % C
		index++
	}
}

func main() {
	var X0, A, B, C int
	// Ввод параметров
	fmt.Scan(&X0, &A, &B, &C)
	generateLCG(X0, A, B, C)
}
